### PR TITLE
API - Documentation chapter

### PR DESCRIPTION
**Pull request by nikajukic**
Task: [#356](https://app.productive.io/1-infinum/tasks/1670946)

#### Aim

The general aim is to rewrite the existing API guide and split it into multiple chapters. One of the chapters should explain documentation.

#### Solution

Add a documentation chapter. Explain why do we need documentation and what information should we put in it.

Written by @lovro-bikic.

### CODE REVIEW

File: API/Documentation.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+- **Overview** - Placed at the top of the documentation, the overview should explain: what the API does, architecture (e.g. REST), format (e.g. JSON:API or GraphQL) and should provide general information (base URL, response types, etc.).
+- **Format requirements** - If you use JSON:API, GraphQL or some other format, they typically come with requirements, like setting `Content-Type`/`Accept` headers, which the clients must abide by.
+- **CORS** - If applicable, document your CORS policy (allowed origins, methods, exposed headers, credentials, etc.).
+- **Authentication** - Applicable if endpoints are authenticated. Explain the authentication method(s) used (API key, cookies, JWT, etc.) and how to obtain credentials.
+- **Authorization** - If the outcome of requests depends on the user's permissions, then that should also be documented. If authorization logic varies between endpoints, each endpoint should expain its logic.
```
Comment (nikajukic) [inline API/Documentation.md:15]: typo: explain

File: API/Documentation.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+- **Diagrams** - Complex flows are hard to explain with just words. If the API, for example, controls a state machine, words will struggle to explain its states and transitions. Diagrams, even simple ones created in tools like [draw.io](https://draw.io), will help clients tremendously in understanding the underlying logic.<br />
+*Tip*: keep raw, editable files in the repository along with exported formats (like PNG). That way other developers can later update existing diagrams instead of having to create new ones.
+
+## Hall of fame
+
+Everything written about documentation practices in the preceding paragraphs sets a basis on which API documentation should be written. But it's just that — a basis. Real-world APIs are big and complex and to write comprehensible documentation for them means to consider many aspects not pointed out here. Instead of trying to put everything into words, here are some API documentations which prove that it's possible to manage complexity and present it in an understandable and user-friendly way to the clients.
+
+- [Stripe API](https://stripe.com/docs/api)
+# TODO: add other examples
```
Comment (vr4b4c) [inline API/Documentation.md:31]: https://developer.productive.io/ 🤘

Comment (d4be4st) [reply API/Documentation.md:31]: https://underline-api.byinfinum.co/api/v1/docs/

File: API/Documentation.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+- **Authorization** - If the outcome of requests depends on the user's permissions, then that should also be documented. If authorization logic varies between endpoints, each endpoint should expain its logic.
+- **Errors** - All APIs are subject to errors. Errors are roadblocks, preventing clients from doing what they need to do. And when a roadblock comes with either no explanation or a vague one (`PC LOAD LETTER`), frustration ensues.<br />
+Document <u>all</u> common errors and supply descriptions. Descriptions should explain what happened and what the client can do to remove the error. Document even the 5xx messages, no matter how obvious they seem.
+- **Changelog** - Major changes to the API should be documented for historical reference. It's an optional step if the API will be used by a limited number of users to whom you can communicate changes directly, though it's useful to have even then.
+- **Endpoints** - The main body of the documentation, each endpoint should: describe what it is, document available HTTP verbs, and show example requests and responses for happy and sad paths. If a response error is specific to an endpoint, document it.
```
Comment (d4be4st) [inline API/Documentation.md:19]: **General params**
We should include how to pagingation/sorting/filtering and other included features like fileld sparsets work.
(basically if API is using JSONAPI::Spec explain how its works :D )

File: API/Documentation.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+- **Endpoints** - The main body of the documentation, each endpoint should: describe what it is, document available HTTP verbs, and show example requests and responses for happy and sad paths. If a response error is specific to an endpoint, document it. 
+
+*Bonus points:*
+- **Quickstart guide** - Onboard clients quickly by providing an example request (e.g. in cURL) which they can copy/paste to test out and modify. This approach is the fastest way to get clients to learn using the API.
+- **Diagrams** - Complex flows are hard to explain with just words. If the API, for example, controls a state machine, words will struggle to explain its states and transitions. Diagrams, even simple ones created in tools like [draw.io](https://draw.io), will help clients tremendously in understanding the underlying logic.<br />
+*Tip*: keep raw, editable files in the repository along with exported formats (like PNG). That way other developers can later update existing diagrams instead of having to create new ones.
+
```
Comment (d4be4st) [inline API/Documentation.md:25]: how about adding a section for "Direct upload flow" if application supports it?
Its a pretty complex flow and i think our frontend devs appreciate it.

Comment (cilim) [reply API/Documentation.md:25]: I'm 99% sure we had this chapter sometimes, somewhere...maybe it got transferred to devproc handbook?

Comment (nikajukic) [reply API/Documentation.md:26]: We have it mentioned [here](https://infinum.com/handbook/books/rails/development-practices/file-uploads#direct-s3-upload)

Comment (cilim) [reply API/Documentation.md:26]: Yeah, I think the one I'm thinking of had pictures, steps and such...but it's lost somewhere... 😅

Comment (d4be4st) [reply API/Documentation.md:26]: In the general section of the underline api documentation :D

File: API/Documentation.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+- **Authentication** - Applicable if endpoints are authenticated. Explain the authentication method(s) used (API key, cookies, JWT, etc.) and how to obtain credentials.
+- **Authorization** - If the outcome of requests depends on the user's permissions, then that should also be documented. If authorization logic varies between endpoints, each endpoint should explain its logic.
+- **Errors** - All APIs are subject to errors. Errors are roadblocks, preventing clients from doing what they need to do. And when a roadblock comes with either no explanation or a vague one (`PC LOAD LETTER`), frustration ensues.<br />
+Document <u>all</u> common errors and supply descriptions. Descriptions should explain what happened and what the client can do to remove the error. Document even the 5xx messages, no matter how obvious they seem.
+- **Changelog** - Major changes to the API should be documented for historical reference. It's an optional step if the API will be used by a limited number of users to whom you can communicate changes directly, though it's useful to have even then.
```
Comment (melcha) [inline API/Documentation.md:18]: Do we have an example of this on a project?

**Review by melcha — APPROVED** [review]

File: API/Documentation.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+*Tip*: keep raw, editable files in the repository along with exported formats (like PNG). That way other developers can later update existing diagrams instead of having to create new ones.
+
+## Hall of fame
+
+Everything written about documentation practices in the preceding paragraphs sets a basis on which API documentation should be written. But it's just that — a basis. Real-world APIs are big and complex and to write comprehensible documentation for them means to consider many aspects not pointed out here. Instead of trying to put everything into words, here are some API documentations which prove that it's possible to manage complexity and present it in an understandable and user-friendly way to the clients.
+
+- [Stripe API](https://stripe.com/docs/api)
+- [Productive](https://developer.productive.io/)
+- [Underline](https://underline-api.byinfinum.co/api/v1/docs/)
```
Comment (d4be4st) [inline API/Documentation.md:33]: you will have to remove underline :( 
its not a public project

but we do have the one from our exapmle app

https://cekila.byinfinum.co/api/v1/docs/

File: API/Documentation.md [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+## Automating documentation
+
+Documentation writing is a labor. Most of it has to be done by hand — nothing can replace that — but some things can be automated. One of the things we can automate are example requests and responses.
+
+### Dox
+
+Infinum's very own gem that does the majority of work in regards to API documentation for the developer. It's super easy to setup and with just a couple of changes to your existing request specs you get an updated API documentation after the specs finish running.
+Check the dox documentation by following [this link](https://github.com/infinum/dox).
```
Comment (d4be4st) [inline API/Documentation.md:42]:
```suggestion
Check the [dox documentation](https://github.com/infinum/dox).
```

File: API/Documentation.md (hunk 5 above)
Comment (d4be4st) [reply API/Documentation.md:18]: On truck... https://api-truck-staging.byinfinum.co/v4/docs/#changelog
Dont know if we can show that as well...

**Review by d4be4st — CHANGES_REQUESTED** [review]

**Review by vr4b4c — APPROVED** [review]

File: API/Documentation.md (hunk 5 above)
Comment (d4be4st) [reply API/Documentation.md:18]: On one project, yes

**Pull request merged**